Token expiry n integration tests

hellais opened the pull request:
This superseeds https://github.com/ooni/orchestra/pull/53 & includes integration testing cherry picked commits from: https://github.com/ooni/orchestra/pull/44

hellais commented:
I think this is good to merge into master.

@bassosimone can you take a look at this if it seems reasonable to you?

hellais commented on orchestrate/orchestrate/sched/scheduler.go:
```diff
@@ -259,17 +259,18 @@ func (j *Job) GetTargets(jDB *JobDB) []*JobTarget {
 	// it.
 	query = fmt.Sprintf("SELECT id, token, platform FROM %s",
 		pq.QuoteIdentifier(common.ActiveProbesTable))
+	query += " WHERE is_token_expired = false AND token != '' AND"
```
This is the key to supporting empty tokens in the DB and tracking the expiration of tokens.

hellais commented on orchestrate/orchestrate/sched/scheduler.go:
```diff
@@ -348,75 +349,120 @@ type NotifyReq struct {
+			errorStr == "NotRegistered" ||
+			errorStr == "InvalidRegistration" {
+			return ErrExpiredToken
+		}
```
I haven't actually tested if these keys are good, but according to the docs I should be using these.

I am thinking that it's probably important to have sentry added to orchestra ASAP to be able to log this sort of stuff.

bassosimone replied:
> it's probably important to have sentry added to orchestra

I concur. However it seems we handle all errors already, so it's probably not `P0` or `P1`.

hellais commented on orchestrate/orchestrate/sched/scheduler.go:
```diff
@@ -348,75 +349,120 @@ type NotifyReq struct {
 
-// TaskNotifyOrchestra this will tell the notify backend to notify the client of
-// the task
-func TaskNotifyOrchestra(bu string, jt *JobTarget) error {
```
This was dead code, so it was removed.

hellais commented on registry/registry/handler/admin.go:
```diff
@@ -141,7 +143,8 @@ func ListClients(db *sqlx.DB, q ClientsQuery) ([]ActiveClient, error) {
-			probe_id FROM %s`,
+			probe_id
+			FROM %s`,
 		pq.QuoteIdentifier(common.ActiveProbesTable))
```
Is it correct to show in the admin view also probes that don't have a valid token?

I think it is reasonable, but maybe I am wrong.

lorenzoPrimi replied:
I guess it is, but we can't send notifications, just check the probe info

bassosimone replied:
@hellais if they're flagged as such, it's probably fine. We should anyway reserve the right to change our mind if having too many clients in such state starts to trouble us, and we cannot do anything about it.

hellais replied:
Currently, since work on the frontend is needed, it will not show them differently. I will file a ticket about this though.

hellais replied:
Filed: https://github.com/ooni/orchestra/issues/61

bassosimone replied:
Thanks

hellais commented on registry/registry/handler/registry.go:
```diff
@@ -168,9 +169,6 @@ func Update(db *sqlx.DB, clientID string, req ClientData) error {
 func Register(db *sqlx.DB, req ClientData) (string, error) {
-	if (req.Platform == "ios" || req.Platform == "android") && req.Token == "" {
-		return "", errors.New("missing device token")
-	}
```
This removed the constraint on the token being not null.

bassosimone replied:
Above you asked whether we should show clients with an empty token. Do I understand the context correctly, if I say that this allows clients using f-droid to register? (I _fail_ to fully understand the PR's motivation, please help)

lorenzoPrimi commented:
This PR closes https://github.com/ooni/orchestra/issues/41 right?

bassosimone commented on orchestrate/orchestrate/sched/scheduler.go:
```diff
@@ -624,16 +645,16 @@ func Notify(jt *JobTarget, jDB *JobDB) error {
-			err = TaskNotifyOrchestra(
-				viper.GetString("core.notify-url"),
-				jt)
-		*/
```
I believe we should change the error to mention that `proteus-notify is no longer supported`

hellais replied:
Ok I can do that. I guess "no longer supported" is stronger than being "deprecated".

bassosimone commented on orchestrate/orchestrate/sched/scheduler.go:
```diff
@@ -624,16 +645,16 @@ func Notify(jt *JobTarget, jDB *JobDB) error {
+		if err != nil {
+			return err
+		}
+	} else if err != nil {
```
Say we've failed to correctly map all relevant error conditions to `ErrExpiredToken`. Then, what is the impact of not expiring a token when we should have done so? What would be the impact of unconditionally expiring the token regardless of the kind of error that occurred?

hellais replied:
That is basically what has been happening until now. The impact is that we will keep re-using that token, hence sending push notifications via APN/firebase and the notification server telling us "hey that token is no good".
I think it's much safer to use an expired token and have the notification be ignored by the push notification server (the "damage" is just that we waste some bandwidth) vs risking to mark a token as expired and not sending a notification to a valid client.

hellais commented on orchestrate/orchestrate/sched/scheduler.go:
```diff
@@ -348,75 +349,120 @@ type NotifyReq struct {
-		},
-	}
+//GoRushLog contains details about the failure. It is available when core->sync
+// is true.
```
It's a settings inside of the gorush configuration file: 
>   sync: false # set true if you need get error message from fail push notification in API response.

https://github.com/appleboy/gorush#features

hellais commented on orchestrate/orchestrate/sched/scheduler.go:
```diff
@@ -348,75 +349,120 @@ type NotifyReq struct {
+
+// Notification payload of a notification to be sent to gorush
+type Notification struct {
+	Type    int // 1 is message 2 is Event
```
This is actually dead code now. I am removing it.

Thanks for spotting it!

hellais commented:
Thanks for the reviews!

I believe I have addressed all the comments and feedback from @bassosimone in the last commit.

If the CI passes and a last smoke test is good, I am going to cut a release and deploy it.
